src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java

```diff
@@ -77,13 +84,70 @@ public void add(
         {
             throw new IllegalStateException("Too many realms");
         }
-        realmsIdsByName.put(realm, 1L << nextRealmBitShift++);
+        realmsIdsByName.put(realm, new OAuthRealm(1L << nextRealmBitShift++));
     }
 
     public long resolve(
-        String realm)
+        String realm,
+        String[] scopes)
     {
-        return realmsIdsByName.getOrDefault(realm, NO_AUTHORIZATION);
+        // If realm doesn't exist, add it to realms
+        if(!realmsIdsByName.containsKey(realm))
+        {
+            add(realm);
+        }
+        final OAuthRealm realmObject = realmsIdsByName.get(realm);
+        if(realmObject == null)
+        {
+            return NO_AUTHORIZATION;
+        }
+        long realmBit = realmObject.realmBit;
+        // if not already there, add the scope to the map, assign each scope a bit
+        // which determines which low bit will be flipped if that scope is present
+        for (int i = 0; i < scopes.length; i++)
+        {
+            final String scope = scopes[i];
+            // check if scope's bit has been set and if scope can be added
+            if(!realmObject.scopeBitAssigned(scope) && !realmObject.supplyScopeBit(scope))
+            {
+                throw new IllegalStateException("Too many scopes");
+            }
+            final long bit = realmObject.getScopeBit(scope);
+            realmBit |= bit;
+        }
+        return realmBit;
+    }
+
+    public long lookup(
+        JsonWebSignature verified)
+    {
+        final String realmName = verified.getKeyIdHeaderValue();
+        try
+        {
+            final JwtClaims claims = JwtClaims.parse(verified.getPayload());
+            final Object scopeClaim = claims.getClaimValue(SCOPES_CLAIM);
+            final String[] roleNames = scopeClaim != null ?
+                    splitScopes(scopeClaim.toString())
+                    : EMPTY_STRING_ARRAY;
+            final OAuthRealm realm = realmsIdsByName.get(realmName);
+            if(realm == null)
+            {
+                return NO_AUTHORIZATION;
+            }
+            long authorizationBits = realm.realmBit;
+            for (int i = 0; i < roleNames.length; i++)
+            {
+                final String scope = roleNames[i];
+                final long bit = realm.getScopeBit(scope);
+                authorizationBits |= bit;
+            }
+            return authorizationBits;
+        }
+        catch (JoseException | InvalidJwtException e)
+        {
+            // TODO: diagnostics?
+            return NO_AUTHORIZATION;
+        }
```

Comment:
Move lines 132-136 first in the method to detect when `realmName` is not recognized, then move the rest of this method to `OAuthRealm.lookup(scopeNames)`.